### ISSUE TITLE
Relay mode: Is it possible to return `ListConnection[Item] | Error`?

### ISSUE THREAD

**Issue by pomidoroshev** [strawberry-graphql/strawberry#2893]
Hello everyone!

I have the following idea:
- if the user is authenticated, he can access the fruits,
- if he is anonymous, he gets an error.

I took the example from the [docs](https://strawberry.rocks/docs/guides/relay), trying to add the error to the response, but it didn't work.

Approach 1: `@relay.connection(relay.ListConnection[Fruit] | Error)`

```python
import strawberry
from strawberry import relay
from strawberry.types import Info
from typing import Iterable

@strawberry.type
class Fruit(relay.Node):
    ...

    @classmethod
    def resolve_nodes(cls, *, info, node_ids, required = False):
        return [
            all_fruits[int(nid)] if required else all_fruits.get(nid)
            for nid in node_ids
        ]

all_fruits = { ... }

@strawberry.type
class Error:
    message: str

@strawberry.type
class Query:
    node: relay.Node = relay.node()

    @relay.connection(relay.ListConnection[Fruit] | Error)
    def fruits(self) -> Iterable[Fruit]:
        # return all the fruits or an error
        ...

schema = strawberry.Schema(query=Query)
```

This one raises `RelayWrongAnnotationError: Wrong annotation used on field "fruits". It should be annotated with a "Connection" subclass.`

Approach 2: `-> Iterable[Fruit] | Error`
```python
...

@strawberry.type
class Query:
    node: relay.Node = relay.node()

    @relay.connection(relay.ListConnection[Fruit])
    def fruits(self) -> Iterable[Fruit] | Error:
        # return all the fruits or an error
        ...
...
```

In this case, a different exception is raises: `TypeError: Query fields cannot be resolved. issubclass() arg 1 must be a class`.

Am I doing it wrong?

By the way, I had similar issues with `strawberry-django-plus`, but they were quite different:
- https://github.com/blb-ventures/strawberry-django-plus/issues/223
- https://github.com/blb-ventures/strawberry-django-plus/issues/225

**Comment by patrick91** [comment]
hi @pomidoroshev!

unfortunately GraphQL doesn't allow unions with lists or scalars, so you'd need to wrap your list inside a GraphQL type:

```python
@strawberry.type
class FruitsResult:
    fruits: List[Fruit]

@strawberry.type
class Query:
    node: relay.Node = relay.node()

    @relay.connection(relay.ListConnection[Fruit])
    def fruits(self) -> FruitsResult | Error:
        # return all the fruits or an error
        ...
```

but I'm not 100% this works with the relay implementation though 😊

**Comment by pomidoroshev** [comment]
@patrick91 Yes, I'm aware about the limitations of the lists. However, `relay.ListConnection` isn't actually a GraphQL list, the generated schema is kinda:

```graphql
# ...

type FruitConnection {
  pageInfo: PageInfo!
  edges: [FruitEdge!]!
}

type Query {
  node(id: GlobalID!): Node!
  fruits(
    before: String = null
    # ...
  ): FruitConnection!
}
```

So there's already a `FruitResult`-like object, which contains `edges` as a list of objects.

**Comment by patrick91** [comment]
oh, right, so you'd like to get a schema like this:

```graphql
type FruitConnection {
  pageInfo: PageInfo!
  edges: [FruitEdge!]!
}

type Query {
  fruits(
    before: String = null
    # ...
  ): FruitConnection! | Error
}
```

right?

**Comment by pomidoroshev** [comment]
@patrick91 Yes, that what I was thinking about. The problem in general is that in my project I have a few queries that can be resolved by anonymous users, and the rest of them are protected. Currently, the protected queries return a connection with empty list of `edges`, which is just an empty list for the client. If there was an error, the client could request a new access token and retry the query.

**Comment by bellini666** [comment]
Hey @pomidoroshev ,

Continuing our discussion from strawberry-django-plus.

Now that the integration is here, we do have a better way to workaround this.

The first thing that comes to my mind here: You should use `@strawberry.field` instead of `@strawberry.connection` here. Why? Because `@strawberry.connection` is supposed to be a shortcut for you to be able to only worry about returning an iterable/generator, and the `Connection` itself handles the pagination for you

Something like this:

```python
@strawberry.field
def fruits(
    self,
    info: Info,
    before: str | None = None,
    after: str | None = None,
    first: int | None = None,
    last: int | None = None,
) -> relay.ListConnection[Fruit] | Error:
    try:
        fruits_iterable = get_cfruits()
    except Exception:
        return error

    return ListConnection.resolve_connection(
        fruits_iterable,
        info=info,
        before=before,
        after=after,
        first=first,
        last=last,
    )
```

@patrick91 what do you think? Would you suggest a different approach? Is there anything that we would need to improve in our side to better support those situations?

**Comment by pomidoroshev** [comment]
@bellini666 Hi, thanks for the workaround. I find it quite odd though, so I'll probably try to implement an abstraction for that, because copy-pasting this code manually across all the queries that I have so far doesn't seem like a proper solution
